prov/gni/src/gnix_msg.c

```diff
@@ -1680,3 +2529,434 @@ ssize_t _gnix_send(struct gnix_fid_ep *ep, uint64_t loc_addr, size_t len,
+	struct gnix_fab_req *req = NULL;
+	struct fid_mr *auto_mr[count];
+
+	if (!(flags & FI_TAGGED)) {
+		if (!ep->ep_ops.msg_send_allowed)
+			return -FI_EOPNOTSUPP;
+	} else {
+		if (!ep->ep_ops.tagged_send_allowed)
+			return -FI_EOPNOTSUPP;
+	}
+
+	ret = _gnix_vc_ep_get_vc(ep, dest_addr, &vc);
+	if (ret != FI_SUCCESS) {
+		return ret;
+	}
+
+	req = _gnix_fr_alloc(ep);
+	if (req == NULL) {
+		return -FI_ENOSPC;
+	}
+
+	GNIX_DEBUG(FI_LOG_EP_DATA, "Created req - %p\n", req);
+
+	/* calculate cumulative size of the iovec buf lens */
+	for (i = 0; i < count; i++) {
+		/* TODO: handle possible overflow */
+		cum_len += iov[i].iov_len;
+	}
+
+	/* Fill out fabric request */
+	if (flags & FI_TAGGED) {
+		req->type = GNIX_FAB_RQ_TSENDV;
+		req->msg.tag = tag;
+		req->msg.ignore = ignore;
+	} else {
+		req->type = GNIX_FAB_RQ_SENDV;
+		req->msg.tag = 0;
+		req->msg.ignore = ~0;
+	}
+
+	req->gnix_ep = ep;
+	req->user_context = context;
+	req->vc = vc;
+	req->work_fn = _gnix_send_req;
+	req->flags = 0; /* Flags that apply to all message types? */
+	req->msg.send_flags = flags | GNIX_MSG_IOV;
+	req->msg.imm = 0;
+
+	/*
+	 * If the cum_len is >= ep->domain->params.msg_rendezvous_thresh
+	 * transfer the iovec entries individually.
+	 *
+	 * For this case, use CtPostFma for iovec lengths that are smaller than
+	 * the rendezvous thresh. For CtPostFma:
+	 * the sum of the iov lens must be either <= 1GB or <= 1MB if the comm
+	 * dom is configured with FmaSharing.
+	 * otherwise use PostRdma.
+	 */
+	if (cum_len >= ep->domain->params.msg_rendezvous_thresh) {
+		req->msg.send_iov_md = malloc(
+			sizeof(struct gnix_fid_mem_desc *) * count);
+		if (!mdesc) {	/* Register the memory for the user */
+			for (i = 0; i < count; i++) {
+				auto_mr[i] = NULL;
+				ret = gnix_mr_reg(&ep->domain->domain_fid.fid,
+						  iov[i].iov_base,
+						  iov[i].iov_len,
+						  FI_READ | FI_WRITE, 0, 0, 0,
+						  auto_mr + i, NULL);
+
+				if (ret != FI_SUCCESS) {
+					GNIX_DEBUG(FI_LOG_EP_DATA,
+						   "Failed to auto-register"
+						   " local buffer: %d\n",
+						   ret);
+					return ret;
+				}
+
+				req->msg.send_iov_md[i] = container_of(
+					(void *) auto_mr[i],
+					struct gnix_fid_mem_desc,
+					mr_fid);
+
+				GNIX_DEBUG(FI_LOG_EP_DATA, "req->msg.send_iov_req->msg.send_iov_md[%d] "
+					   "= %p\n", i,
+					   req->msg.send_iov_md[i]);
+
+				GNIX_DEBUG(FI_LOG_EP_DATA, "auto-reg MR: %p\n",
+					   req->msg.send_iov_md[i]);
+
+			}
+
+			req->msg.send_flags |= FI_LOCAL_MR;
+		} else {	/* User registered their memory */
+			for (i = 0; i < count; i++) {
+				req->msg.send_iov_md[i] =
+					container_of(mdesc[i],
+						     struct gnix_fid_mem_desc,
+						     mr_fid);
+			}
+		}
+
+		/* The default iov_limit for the gni provider keeps the
+		 * data portion of the control packet near the rndzv threshold
+		 * however, the hdr is not accounted for.
+		 * TODO: Account for the header size in the iov_limit.
+		 */
+		req->msg.send_addr = (uint64_t) iov;
```

Comment:
Are you sure we're allowed to reference the user's iov buffer after returning from fi_sendv?